gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaAvroSchemaRegistry.java

```diff
@@ -0,0 +1,289 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.source.extractor.extract.kafka;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.avro.Schema;
+import org.apache.commons.httpclient.Header;
+import org.apache.commons.httpclient.HttpClient;
+import org.apache.commons.httpclient.HttpException;
+import org.apache.commons.httpclient.HttpStatus;
+import org.apache.commons.httpclient.MultiThreadedHttpConnectionManager;
+import org.apache.commons.httpclient.URIException;
+import org.apache.commons.httpclient.methods.GetMethod;
+import org.apache.commons.httpclient.methods.PostMethod;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.Maps;
+
+
+/**
+ * A schema registry class that provides two services: get the latest schema of a topic, and register a schema.
+ *
+ * @author ziliu
+ *
+ */
+public class KafkaAvroSchemaRegistry {
+
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaAvroSchemaRegistry.class);
+
+  private static final int GET_SCHEMA_BY_ID_MAX_TIRES = 3;
+  private static final int GET_SCHEMA_BY_ID_MIN_INTERVAL_SECONDS = 1;
+  private static final String GET_RESOURCE_BY_ID = "/id=";
+  private static final String GET_RESOURCE_BY_TYPE = "/latest_with_type=";
+  private static final String SCHEMA_ID_HEADER_NAME = "Location";
+  private static final String SCHEMA_ID_HEADER_PREFIX = "/id=";
+
+  private final LoadingCache<String, Schema> cachedSchemasById;
+
+  private final HttpClient httpClient;
+
+  private final String url;
+
+  public KafkaAvroSchemaRegistry(String url) {
+    this.url = url;
+    this.cachedSchemasById =
+        CacheBuilder.newBuilder().maximumSize(1000).expireAfterWrite(10, TimeUnit.MINUTES)
+            .build(new KafkaSchemaCacheLoader());
+    this.httpClient = new HttpClient(new MultiThreadedHttpConnectionManager());
+  }
+
+  /**
+   * Get schema from schema registry by ID
+   *
+   * @param id Schema ID
+   * @return Schema with the corresponding ID
+   * @throws SchemaNotFoundException if ID not found
+   */
+  public synchronized Schema getSchemaById(String id) throws SchemaNotFoundException {
+    try {
+      return cachedSchemasById.get(id);
+    } catch (ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Get the latest schema of a topic.
+   *
+   * @param topic topic name
+   * @return the latest schema
+   * @throws SchemaNotFoundException if topic name not found.
+   */
+  public synchronized Schema getLatestSchemaByTopic(String topic) throws SchemaNotFoundException {
```

Comment:
It does not seem this method is accessing any shared mutable objects? If yes, then there's no need to use `synchronized`.